### PR TITLE
Large refactors

### PR DESCRIPTION
**Pull request by erikh**
These are some large refactors that establish a runner framework suitable for creating new runners. This framework lives under the `fw/fw.go` hieriarchy of types and imports, and supporting code lives under `fw/` for the whole thing.

### CODE REVIEW

**Comment by erikh** [comment]
@capoferro @unclejack this is ready for review; please do it today if you get time, this needs to move forward asap.

**Comment by erikh** [comment]
@unclejack @capoferro PTAL

File: runners/overlay-runner/runner.go [hunk 1]
```diff
@@ -1,71 +1,95 @@
...
-// the log.
-func (r *Run) StartLogger(rc io.Reader) {
-	go func() {
-		if err := r.Config.Clients.Asset.Write(r.QueueItem.Run.ID, rc); err != nil {
-			r.Logger.Error(err.Wrapf("Writing log for Run ID %d", r.QueueItem.Run.ID))
-		}
-	}()
+// BeforeRun is executed before the next run is started.
+func (r *Runner) BeforeRun(ctx *fw.Context) *errors.Error {
+	r.CurrentRun = NewRun(ctx.RunCtx, ctx.RunCancelFunc, ctx.QueueItem, r.Config, r.LogsvcClient(ctx))
```
Comment (capoferro) [inline runners/overlay-runner/runner.go]: This line seems like something that the framework should do automatically. Will each runner need to set up `CurrentRun` in their `BeforeRun`?

Comment (erikh) [reply runners/overlay-runner/runner.go]: this is a implementation specific bit; not related to the framework.

**Pull request merged**